src/GuzzleRequestAdapter.php

```diff
@@ -281,152 +356,282 @@ public function getBaseUrl(): string
+
+            $rootParseNode = $this->getRootParseNode($response, $errorSpan);
+            if ($errorClass !== null) {
+                $spanForDeserialization = $this->tracer->spanBuilder('ParseNode.GetObjectValue()')
+                    ->setParent(Context::getCurrent())
+                    ->addLink($errorSpan->getContext())
+                    ->startSpan();
+                $error                  = $rootParseNode->getObjectValue($errorClass);
+                $this->setResponseType($errorClass[0], $spanForDeserialization);
+                $spanForDeserialization->end();
+
+            } else {
+                $error = null;
+            }
 
-        $rootParseNode = $this->getRootParseNode($response);
-        if ($errorClass !== null) {
-            $error = $rootParseNode->getObjectValue($errorClass);
-        } else {
-            $error = null;
-        }
-        if ($error && is_subclass_of($error, ApiException::class)) {
-            $error->setResponseStatusCode($response->getStatusCode());
-            $error->setResponseHeaders($response->getHeaders());
-            throw $error;
+            if ($error && is_subclass_of($error, ApiException::class)) {
+                $span->setAttribute(self::ERROR_BODY_FOUND_ATTRIBUTE_NAME, true);
+                $error->setResponseStatusCode($response->getStatusCode());
+                $error->setResponseHeaders($response->getHeaders());
+                $errorSpan->recordException($error, ['know_error' => true, 'message' => $responseBodyContents]);
+                throw $error;
+            }
+            $otherwise = new ApiException("Unsupported error type " . get_debug_type($error));
+            $otherwise->setResponseStatusCode($response->getStatusCode());
+            $otherwise->setResponseHeaders($response->getHeaders());
+            $errorSpan->recordException($otherwise, ['known_error' => false, 'message' => $responseBodyContents]);
+            throw $otherwise;
+        } finally {
+            $scope->detach();
+            $errorSpan->end();
+            $span->end();
         }
-        $otherwise = new ApiException("Unsupported error type ". get_debug_type($error));
-        $otherwise->setResponseStatusCode($response->getStatusCode());
-        $otherwise->setResponseHeaders($response->getHeaders());
-        throw $otherwise;
     }
 
     /**
      * @param ResponseInterface $response
      * @return bool
      */
-    private function is204NoContentResponse(ResponseInterface $response): bool{
+    private function is204NoContentResponse(ResponseInterface $response): bool
+    {
         return $response->getStatusCode() === 204;
     }
+    private const QUERY_PARAMETERS_CLEANUP_REGEX = "/\{+?[^}]+}/";
+
+    private function startTracingSpan(RequestInformation $requestInfo, string $methodName): SpanInterface
+    {
```

Comment:
We should check for any ObservabilityOptions in the request info->options so that we override the configuration across the entire package.